packages/grid/x-data-grid/src/utils/createSelector.ts

```diff
@@ -48,12 +48,99 @@ const missingInstanceIdWarning = buildWarning([
 ]);
 
-export const createSelector: CreateSelectorFunction = (...args: any) => {
+function isAPIRef(value: any) {
+  return 'current' in value && 'instanceId' in value.current;
+}
+
+const DEFAULT_INSTANCE_ID = { id: 'default' };
+
+export const createSelector = ((
+  a: Function,
+  b?: Function,
+  c?: Function,
+  d?: Function,
+  e?: Function,
+  f?: Function,
+  ...rest: any[]
+) => {
+  if (rest.length > 0) {
+    throw new Error('Unsupported number of selectors');
+  }
+
+  let selector: any;
+
+  if (a && b && c && d && e && f) {
+    selector = (stateOrApiRef: any, instanceIdParam: any) => {
+      const isAPI = isAPIRef(stateOrApiRef);
+      const instanceId =
+        instanceIdParam ?? (isAPI ? stateOrApiRef.current.instanceId : DEFAULT_INSTANCE_ID);
+      const state = isAPI ? stateOrApiRef.current.state : stateOrApiRef;
+      const va = a(state, instanceId);
+      const vb = b(state, instanceId);
+      const vc = c(state, instanceId);
+      const vd = d(state, instanceId);
+      const ve = e(state, instanceId);
+      return f(va, vb, vc, vd, ve);
+    };
+  } else if (a && b && c && d && e) {
+    selector = (stateOrApiRef: any, instanceIdParam: any) => {
+      const isAPI = isAPIRef(stateOrApiRef);
+      const instanceId =
+        instanceIdParam ?? (isAPI ? stateOrApiRef.current.instanceId : DEFAULT_INSTANCE_ID);
+      const state = isAPI ? stateOrApiRef.current.state : stateOrApiRef;
+      const va = a(state, instanceId);
+      const vb = b(state, instanceId);
+      const vc = c(state, instanceId);
+      const vd = d(state, instanceId);
+      return e(va, vb, vc, vd);
+    };
+  } else if (a && b && c && d) {
+    selector = (stateOrApiRef: any, instanceIdParam: any) => {
+      const isAPI = isAPIRef(stateOrApiRef);
+      const instanceId =
+        instanceIdParam ?? (isAPI ? stateOrApiRef.current.instanceId : DEFAULT_INSTANCE_ID);
+      const state = isAPI ? stateOrApiRef.current.state : stateOrApiRef;
+      const va = a(state, instanceId);
+      const vb = b(state, instanceId);
+      const vc = c(state, instanceId);
+      return d(va, vb, vc);
+    };
+  } else if (a && b && c) {
+    selector = (stateOrApiRef: any, instanceIdParam: any) => {
+      const isAPI = isAPIRef(stateOrApiRef);
+      const instanceId =
+        instanceIdParam ?? (isAPI ? stateOrApiRef.current.instanceId : DEFAULT_INSTANCE_ID);
+      const state = isAPI ? stateOrApiRef.current.state : stateOrApiRef;
+      const va = a(state, instanceId);
+      const vb = b(state, instanceId);
+      return c(va, vb);
+    };
+  } else if (a && b) {
+    selector = (stateOrApiRef: any, instanceIdParam: any) => {
+      const isAPI = isAPIRef(stateOrApiRef);
+      const instanceId =
+        instanceIdParam ?? (isAPI ? stateOrApiRef.current.instanceId : DEFAULT_INSTANCE_ID);
+      const state = isAPI ? stateOrApiRef.current.state : stateOrApiRef;
+      const va = a(state, instanceId);
```

Comment:
Why do we pass `instanceId` here?
As far as I can tell, the `instanceId` is only being used in `createSelectorMemoized` to get the right cache.

Reply:
One of the sub-selectors might be a memoized selector.